src/engine/LocalVocab.h

```diff
@@ -0,0 +1,69 @@
+// Copyright 2022, University of Freiburg
+// Chair of Algorithms and Data Structures
+// Author: Hannah Bast <bast@cs.uni-freiburg.de>
+
+#pragma once
+
+#include "absl/container/node_hash_map.h"
+
+// A class for maintaing a local vocabulary with contiguous (local) IDs. This is
+// meant for words that are not part of the normal vocabulary (constructed from
+// the input data at indexing time).
+//
+// TODO: This is a first version of this class with basic functionality. Note
+// that the local vocabulary used to be a simple `std::vector<std::string>`
+// defined inside of the `ResultTable` class. You gotta start somewhere.
+class LocalVocab {
+ public:
+  // Create a new, empty local vocabulary.
+  LocalVocab() = default;
+
+  // Prevent accidental copying of a local vocabulary (it can be quite large),
+  // but moving it is OK.
+  //
+  // TODO: does the default move do the "right" thing, that is, move the hash
+  // map instead of copying it?
+  LocalVocab(const LocalVocab&) = delete;
+  LocalVocab(LocalVocab&&) = default;
+
+  // Get ID of a word in the local vocabulary. If the word was already
+  // contained, return the already existing ID. If the word was not yet
+  // contained, add it, and return the new ID.
+  Id getIdAndAddIfNotContained(const std::string& word);
+
+  // The number of words in the vocabulary.
+  size_t size() const { return idsToWordsMap_.size(); }
+
+  // Return true if and only if the local vocabulary is empty.
+  bool empty() const { return idsToWordsMap_.empty(); }
+
+  // Return a const reference to the word.
+  const std::string& getWord(LocalVocabIndex localVocabIndex) const;
+
+  // Merge two local vocabularies if at least one of them is empty. If both are
+  // non-empty, throws an exception.
+  //
+  // TODO: Eventually, we want to have one local vocab for the whole query to
+  // which each operation writes (one after the other). Then we don't need a
+  // merge function anymore.
+  static std::shared_ptr<LocalVocab> mergeLocalVocabsIfOneIsEmpty(
+      std::shared_ptr<LocalVocab> localVocab1,
+      std::shared_ptr<LocalVocab> localVocab2);
```

Comment:
This always copies both `shared_ptrs` although we need at most one of them.
Maybe here (where "creating all overloads" is not necessary because the performance is not that crucial)
the solution is to take a `const shared_ptr&` for both arguments,
especially since we practically never move the arguments in in our actual use cases.